filebeat/input/filestream/input.go

```diff
@@ -231,6 +246,30 @@ func (inp *filestream) open(log *logp.Logger, canceler input.Canceler, path stri
 	return r, nil
 }
 
+func (inp *filestream) startMonitoring(log *logp.Logger, canceler input.Canceler, srcID, path string, f *os.File) {
+	// if there was a previously running monitoring routine, stop it
+	inp.metrics.remove(srcID)
+
+	monitorCtx, monitorCancel := ctxtool.WithFunc(canceler, func() {
+		log.Debug("Stop monitoring of opened file")
+	})
+
+	m := newProgressMetrics(inp.ID, path, srcID, monitorCancel, true)
+	inp.metrics.add(srcID, m)
+	go func() {
+		m := m
+		timed.Periodic(monitorCtx, 30*time.Second, func() error {
+			fi, err := f.Stat()
```

Comment:
The `(*filestream).Run` method will close `f` on shutdown. This go-routine should be guaranteed to be stopped before `Run` closes the file.